codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java

```diff
@@ -266,39 +273,41 @@ public static void addHttp2ToHttpHeaders(int streamId, Http2Headers sourceHeader
 
     /**
      * Converts the given HTTP/1.x headers into HTTP/2 headers.
+     * The following headers are only used if they can not be found in from the {@code HOST} header or the
+     * {@code Request-Line} as defined by <a href="https://tools.ietf.org/html/rfc7230">rfc7230</a>
+     * <ul>
+     * <li>{@link ExtensionHeaderNames#AUTHORITY}</li>
+     * <li>{@link ExtensionHeaderNames#SCHEME}</li>
+     * </ul>
+     * {@link ExtensionHeaderNames#PATH} is ignored and instead extracted from the {@code Request-Line}.
      */
     public static Http2Headers toHttp2Headers(HttpMessage in) throws Exception {
         final Http2Headers out = new DefaultHttp2Headers();
         HttpHeaders inHeaders = in.headers();
         if (in instanceof HttpRequest) {
             HttpRequest request = (HttpRequest) in;
-            out.path(new AsciiString(request.uri()));
-            out.method(new AsciiString(request.method().toString()));
-
-            String value = inHeaders.getAsString(HttpHeaderNames.HOST);
-            if (value != null) {
-                URI hostUri = URI.create(value);
-                // The authority MUST NOT include the deprecated "userinfo" subcomponent
-                value = hostUri.getAuthority();
-                if (value != null) {
-                    out.authority(new AsciiString(AUTHORITY_REPLACEMENT_PATTERN.matcher(value).replaceFirst("")));
+            URI requestTargetUri = URI.create(request.uri());
+            out.path(toHttp2Path(requestTargetUri));
+            out.method(request.method().name());
+
+            // Attempt to take from HOST header before taking from the request-line
+            String host = inHeaders.getAsString(HttpHeaderNames.HOST);
+            boolean shouldSetAuthroity = !isOriginForm(requestTargetUri) && !isAsteriskForm(requestTargetUri);
+            if (host == null) {
+                if (shouldSetAuthroity) {
+                    setHttp2Authority(inHeaders, requestTargetUri, out);
                 }
-                value = hostUri.getScheme();
-                if (value != null) {
-                    out.scheme(new AsciiString(value));
+                setHttp2Scheme(inHeaders, requestTargetUri, true, out);
+            } else {
+                URI hostUri = URI.create(host);
```

Comment:
@trustin - The bracket approach is called out in the standard:

https://tools.ietf.org/html/rfc7230#section-2.7.1 -> https://tools.ietf.org/html/rfc3986#section-3.2.2

```
A host identified by an Internet Protocol literal address, version 6
   [RFC3513] or later, is distinguished by enclosing the IP literal
   within square brackets ("[" and "]").  This is the only place where
   square bracket characters are allowed in the URI syntax.  In
   anticipation of future, as-yet-undefined IP literal address formats,
   an implementation may use an optional version flag to indicate such a
   format explicitly rather than rely on heuristic determination.

      IP-literal = "[" ( IPv6address / IPvFuture  ) "]"

      IPvFuture  = "v" 1*HEXDIG "." 1*( unreserved / sub-delims / ":" )
```

I'll add a unit test for this.